project/src/main/ui/chat/chatscript-parser.gd

```diff
@@ -0,0 +1,378 @@
+
+Mostly contains logic for transitioning into other states.
+"""
+class DefaultState extends AbstractState:
+	var dialog_state: AbstractState
+	
+	func _init(init_chat_tree: ChatTree).(init_chat_tree) -> void:
+		pass
+	
+	
+	func line(line: String) -> String:
+		var result := ""
+		if not line:
+			pass
+		elif line.begins_with("{") and line.ends_with("}"):
+			# json dictionary; set metadata
+			var parsed = parse_json(line)
+			if typeof(parsed) == TYPE_DICTIONARY:
+				chat_tree.meta = parse_json(line)
+				if "version" in chat_tree.meta:
+					chat_tree.meta.erase("version")
+			else:
+				push_warning("Malformed chatscript JSON: %s" % [line])
+		elif line.begins_with("[") and line.ends_with("]"):
+			# state name; update state
+			result = line.trim_prefix("[").trim_suffix("]")
+		else:
+			# unrecognized text; parse as dialog and change state
+			result = dialog_state.line(line)
+			result = StringUtils.default_if_empty(result, DIALOG)
+		return result
+
+
+"""
+Parser state for parsing location data (where a conversation takes place).
+"""
+class LocationState extends AbstractState:
+	func _init(init_chat_tree: ChatTree).(init_chat_tree) -> void:
+			pass
+	
+	
+	"""
+	Syntax:
+		[location]
+		indoors
+	"""
+	func line(line: String) -> String:
+		var result := ""
+		if line:
+			chat_tree.location_id = line
+		else:
+			result = DEFAULT
+		return result
+
+
+"""
+Parser state for parsing character data (participants in a conversation).
+"""
+class CharactersState extends AbstractState:
+	var _character_aliases: Dictionary
+	
+	func _init(init_chat_tree: ChatTree, init_dialog_aliases: Dictionary).(init_chat_tree) -> void:
+		_character_aliases = init_dialog_aliases
+	
+	
+	func line(line: String) -> String:
+		var result := ""
+		if line:
+			_parse_character_name(line)
+		else:
+			result = DEFAULT
+		return result
+	
+	
+	"""
+	Syntax:
+		skins, s, kitchen-9        - a character named 'skins' with an alias 's' spawns at kitchen-9
+		skins, s                   - a character named 'skins' with an alias 's'
+		skins                      - a character named 'skins'
+	"""
+	func _parse_character_name(line: String) -> void:
+		var line_parts := line.split(",")
+		var character_name := "" if line_parts.size() < 1 else line_parts[0].strip_edges()
+		if character_name in ["player", "sensei"]:
+			character_name = "#%s#" % [character_name]
+		var character_alias := "" if line_parts.size() < 2 else line_parts[1].strip_edges()
+		var character_location := "" if line_parts.size() < 3 else line_parts[2].strip_edges()
+		
+		if character_name and character_location:
+			chat_tree.spawn_locations[character_name] = character_location
+		if character_name and character_alias:
+			_character_aliases[character_alias] = character_name
+
+
+"""
+Parser state for parsing dialog data (chat events and branches)
+"""
+class DialogState extends AbstractState:
+	var _character_aliases: Dictionary
+	var _event: ChatEvent
+	var _branch_key := ""
```

Comment:
Document these fields

Reply:
Added comment, `Creature aliases used in chatscript dialog. DialogState references this dictionary when parsing dialog`